src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -1147,6 +1148,77 @@ public static void Fill<T>(ref T refData, nuint numElements, T value)
             return -1; // not found
         }
 
+        public static int IndexOfAnyExcept<T>(ref T searchSpace, T value0, int length)
+        {
+            Debug.Assert(length >= 0, "Expected non-negative length");
+
+            for (int i = 0; i < length; i++)
+            {
+                if (!EqualityComparer<T>.Default.Equals(Unsafe.Add(ref searchSpace, i), value0))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int IndexOfAnyExceptValueType<T>(ref T searchSpace, T value0, int length) where T : struct, IEquatable<T>
+        {
+            Debug.Assert(length >= 0, "Expected non-negative length");
+            Debug.Assert(value0 is byte or short or int or long, "Expected caller to normalize to one of these types");
+
+            if (!Vector128.IsHardwareAccelerated || length < Vector128<T>.Count)
+            {
+                for (int i = 0; i < length; i++)
+                {
+                    if (!Unsafe.Add(ref searchSpace, i).Equals(value0))
+                    {
+                        return i;
+                    }
+                }
+            }
+            else
+            {
+                Vector128<T> notEquals, value0Vector = Vector128.Create(value0);
+                ref T current = ref searchSpace;
+                ref T oneVectorAwayFromEnd = ref Unsafe.Add(ref searchSpace, length - Vector128<T>.Count);
+
+                // Loop until either we've finished all elements or there's less than a vector's-worth remaining.
+                do
+                {
+                    notEquals = ~Vector128.Equals(value0Vector, Vector128.LoadUnsafe(ref current));
+                    if (notEquals != Vector128<T>.Zero)
+                    {
+                        return ComputeIndex(ref searchSpace, ref current, notEquals);
+                    }
+
+                    current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+                }
+                while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd));
+
+                // If any elements remain, process the last vector in the search space.
+                if ((uint)length % Vector128<T>.Count != 0)
+                {
+                    notEquals = ~Vector128.Equals(value0Vector, Vector128.LoadUnsafe(ref oneVectorAwayFromEnd));
+                    if (notEquals != Vector128<T>.Zero)
+                    {
+                        return ComputeIndex(ref searchSpace, ref oneVectorAwayFromEnd, notEquals);
+                    }
+                }
```

Comment:
Do I read this algorithm correctly that for length == Vector128<T>.Count we run both main loop and the block for trailing elements?

Reply:
The trailing loop is guarded by `(uint)length % Vector128<T>.Count != 0`

Reply:
Ah, right